rfcs/0004-transaction.md

```diff
@@ -0,0 +1,291 @@
+
+1. Adding a **conflict check before writing a WriteBatch** to guarantee the transaction isolation.
+2. Adding an **iterator over WriteBatch & Snapshot** to make MVCC possible.
+
+This is also true for the Badger's implementation.
+
+Let's discuss the prerequisites in this section, and evaluate the possible approaches we'd take in SlateDB to smooth the implementation of the transaction feature.
+
+#### Snapshot
+
+LevelDB and its derivatives mostly tags a sequence number to each key, and each Snapshot simply bounds with a sequence number. All the read in the Snapshot filters out the values with bigger sequence number, including a new value or a tombstone on the same key.
+
+At the moment of writing, SlateDB still haven't have the equivant of sequence number yet, instead, we could regard a manifest + wal number to represent a consistent view of a table.
+
+There're an in-progress RFC on exploring the design on Snapshot. We can refer to that RFC for the details later. However, we'd **assume that each key has a sequence number** in this doc, because it's considered as a common practice in LSM-tree based storage engines.
+
+#### Write Batch
+
+This is a sample [code snippet](https://rust.velas.com/rocksdb/struct.WriteBatch.html) of `WriteBatch` in the `rocksdb` crate:
+
+```
+use rocksdb::{DB, Options, WriteBatch};
+
+let path = "_path_for_rocksdb_storage1";
+{
+    let db = DB::open_default(path).unwrap();
+    let mut batch = WriteBatch::default();
+    batch.put(b"my key", b"my value");
+    batch.put(b"key2", b"value2");
+    batch.put(b"key3", b"value3");
+    db.write(batch); // Atomically commits the batch
+}
+```
+
+You may notice that the `WriteBatch` API is very similar to the `Transaction` API we proposed above, the write operation from `WriteBatch` is also considered as atomic. The only differences are:
+
+1. the `WriteBatch` is committed by `db.write(batch)`, while the `Transaction` is committed by `txn.commit()`
+2. beside the write operations, the `Transaction` should also support the read operations, and if one key is updated in the `Transaction`, the read operation should return the updated value during the transaction.
+
+The `WriteBatch` is considered as a prerequisite for the transaction feature, as it plays as a place to buffer the writes, and it makes MVCC possible with an iterator over it in front of the iterator over Snapshot. So we should implement the `WriteBatch` first before the transaction feature.
+
+We can refer to the WriteBatch RFC or PR for the details later.
+
+## Conflict Checking: Snapshot Isolation
+
+The conflict checking in the SI transaction is simple:
+
+1. Track the sequence number when transaction is started in the `TransactionState` struct.
+2. Track the modified keys in the `TransactionState` struct.
+3. When `commit()` is called, check whether the modified keys are modified by others after the transaction started.
```

Comment:
Would the check for modified keys, and subsequent `commit` lock the DB? i.e. There won't be other writes allowed on the memtable after the transaction check started before commit is called.

Reply:
yeah, this discussion is also related to this topic: https://github.com/slatedb/slatedb/pull/260#discussion_r1815785697